runtime/commonTest/src/kotlinx/serialization/SerialDescriptorSpecificationTest.kt

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.CompositeDecoder.Companion.UNKNOWN_NAME
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.test.*
+import kotlin.test.*
+
+class SerialDescriptorSpecificationTest {
+
+    @Serializable
+    class Holder(val a: Int?, @SerialId(42) val b: String = "", @SerialName("c") val d: Long)
+
+    @Test
+    fun testAutoGeneratedDescriptorContract() {
+        testHolderDescriptor(Holder.serializer().descriptor)
+    }
+
+    @Test
+    fun testManuallyConstructedDescriptor() {
+        testHolderDescriptor(HolderDescriptor)
+    }
+
+    private object HolderDescriptor : SerialClassDescImpl(
+        "kotlinx.serialization.SerialDescriptorSpecificationTest.Holder"
+    ) {
+        init {
+            addElement("a")
+            pushDescriptor(IntSerializer.nullable.descriptor)
+
+            addElement("b", true)
+            pushDescriptor(StringSerializer.descriptor)
+            // Can't create an annotation manually
+            pushAnnotation(Holder.serializer().descriptor.findAnnotation<SerialId>(1)!!)
+
+            addElement("c")
+            pushDescriptor(LongSerializer.descriptor)
+        }
+    }
+
+    private fun testHolderDescriptor(d: SerialDescriptor) {
+        assertEquals(StructureKind.CLASS, d.kind)
+        assertEquals("kotlinx.serialization.SerialDescriptorSpecificationTest.Holder", d.serialName)
+        assertEquals(3, d.elementsCount)
+        // Indices
+        run {
+            assertEquals(0, d.getElementIndex("a"))
+            assertEquals(1, d.getElementIndex("b"))
+            assertEquals(2, d.getElementIndex("c"))
+            assertEquals(UNKNOWN_NAME, d.getElementIndex("?"))
+            assertEquals(UNKNOWN_NAME, d.getElementIndex("d"))
+
+            assertEquals("a", d.getElementName(0))
+            assertEquals("b", d.getElementName(1))
+            assertEquals("c", d.getElementName(2))
+        }
+        // Children annotations
+        run {
+            assertEquals(0, d.getElementAnnotations(0).size)
+            d.assertSingleAnnotation(1) { it is SerialId && it.id == 42 }
+            assertEquals(0, d.getElementAnnotations(0).size)
+        }
+        // Optionality
+        run {
```

Comment:
There are too much assertions for a single test run. Do you think it is possible to split this to multiple `@Test`s?

Reply:
I don't think so, the est is pretty isolated, consistent and tests the integrity of the single entity